network/p2p/network.go

```diff
@@ -0,0 +1,169 @@
+func WithValidatorSampling(validators *Validators) ClientOption {
+	return clientOptionFunc(func(options *clientOptions) {
+		options.nodeSampler = validators
+	})
+}
+
+// clientOptions holds client-configurable values
+type clientOptions struct {
+	// nodeSampler is used to select nodes to route Client.AppRequestAny to
+	nodeSampler NodeSampler
+}
+
+// NewNetwork returns an instance of Network
+func NewNetwork(
+	log logging.Logger,
+	sender common.AppSender,
+	metrics prometheus.Registerer,
+	namespace string,
+) *Network {
+	peers := &Peers{}
+	clientDefaults := &clientOptions{
+		nodeSampler: peers,
+	}
+
+	return &Network{
+		Peers:     peers,
+		log:       log,
+		sender:    sender,
+		metrics:   metrics,
+		namespace: namespace,
+		router:    newRouter(log, sender, metrics, namespace, clientDefaults),
+	}
+}
+
+// Network maintains state of the peer-to-peer network and any in-flight
+// requests
+type Network struct {
+	Peers *Peers
+
+	log       logging.Logger
+	sender    common.AppSender
+	metrics   prometheus.Registerer
+	namespace string
+
+	router *router
+}
+
+func (n *Network) AppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, deadline time.Time, request []byte) error {
+	return n.router.AppRequest(ctx, nodeID, requestID, deadline, request)
+}
+
+func (n *Network) AppResponse(ctx context.Context, nodeID ids.NodeID, requestID uint32, response []byte) error {
+	return n.router.AppResponse(ctx, nodeID, requestID, response)
+}
+
+func (n *Network) AppRequestFailed(ctx context.Context, nodeID ids.NodeID, requestID uint32) error {
+	return n.router.AppRequestFailed(ctx, nodeID, requestID)
+}
+
+func (n *Network) AppGossip(ctx context.Context, nodeID ids.NodeID, msg []byte) error {
+	return n.router.AppGossip(ctx, nodeID, msg)
+}
+
+func (n *Network) CrossChainAppRequest(ctx context.Context, chainID ids.ID, requestID uint32, deadline time.Time, request []byte) error {
+	return n.router.CrossChainAppRequest(ctx, chainID, requestID, deadline, request)
+}
+
+func (n *Network) CrossChainAppResponse(ctx context.Context, chainID ids.ID, requestID uint32, response []byte) error {
+	return n.router.CrossChainAppResponse(ctx, chainID, requestID, response)
+}
+
+func (n *Network) CrossChainAppRequestFailed(ctx context.Context, chainID ids.ID, requestID uint32) error {
+	return n.router.CrossChainAppRequestFailed(ctx, chainID, requestID)
+}
+
+func (n *Network) Connected(_ context.Context, nodeID ids.NodeID, _ *version.Application) error {
+	n.Peers.add(nodeID)
+	return nil
+}
+
+func (n *Network) Disconnected(_ context.Context, nodeID ids.NodeID) error {
+	n.Peers.remove(nodeID)
+	return nil
+}
+
+func (n *Network) NewAppProtocol(handlerID uint64, handler Handler, options ...ClientOption) (*Client, error) {
+	client, err := n.router.newAppProtocol(handlerID, handler)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, option := range options {
+		option.apply(client.options)
+	}
```

Comment:
afaict `client.options` is currently shared for all `Client`s on this `Network`... this is probably unexpected behavior (+racy)

Reply:
Yeah that's not intended... will fix.